### PR TITLE
Feature/kah 10

### PR DESCRIPTION
**Pull request by DavidIzaac**

### CODE REVIEW

File: server/api/tests/project.spec.js [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+let mapId;
+
+
+
+setupDB('test');
+
+const testDataManager = initTestDataManager(ProjectModel);
+
+function createMap(projectId,index,mapName){
+    return new Promise((resolve,reject) => {
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: There's no need to wrap supertest requests in promises because they are promises on their own. Just return the request()...

File: server/api/tests/project.spec.js [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+        it(`should respond with 'OK'`, function (done) {
+            let projectName = fixedProjects[1].name;
+            let projectId = testDataManager.getProjectIdByName(projectName)
+            try{
+                request(app)
+                .delete(`/api/projects/${projectId}/delete`)
+                .expect(200)
+                .then(res => {
+                    testDataManager.deleteCollection(projectId)
+                    expect('OK');
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: Why is this here? Can't you put inside a request chain of methods?

File: server/api/tests/project.spec.js [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+                throw err;
+            }
+        });
+    });
+
+    describe(`DELETE /:jobId/ `, () => {
+        it(`should respond with 'OK'`, function (done) {
+            let projectName = fixedProjects[1].name;
+            let projectId = testDataManager.getProjectIdByName(projectName)
+            try{
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: No need for try / catch as you do not use async/await. Instead, please return the promise coming from request(...) invocation.

File: server/api/tests/project.spec.js [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+            }
+            catch(err){
+                console.log(err.message);
+                throw err;
+            }
+        });
+    });
+
+    describe(`DELETE /:jobId/ `, () => {
+        it(`should respond with 'OK'`, function (done) {
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: We are using es6+, switch function for lambda expression (arrow function).

File: server/api/tests/project.spec.js [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+            let projectName = fixedProjects[1].name;
+            let projectId = testDataManager.getProjectIdByName(projectName)
+            try{
+                request(app)
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: Do not use server's instance in supertest, we agreed on using string with url.

File: server/api/tests/project.spec.js [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+    //             console.log(err.message);
+    //             throw err;
+    //         }
+    //     });
+    // });
+
+   
+
+    describe(`GET /:projectId/ `, () => {
+        it(`should respond with the recents maps of the project`, async function (done) {
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: Again, arrow function.

File: server/api/tests/project.spec.js [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+            let projectName = fixedProjects[1].name;
+            let projectId = testDataManager.getProjectIdByName(projectName)
+            let mapName = 'map 1'
+            await createMap(projectId,testDataManager.collection.findIndex(item => item._id == projectId),mapName)
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: This is a part of test's setup - you should put it in beforeEach/beforeAll. Using it here falsifies results of the test by making it run longer and therefore falsifies the performance output.

File: server/api/tests/project.spec.js [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+            try{
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: If you move createMap to beforeAll/beforeEach then there's no need for async/await. Please, remove try/catch once you deal with the above.

File: server/api/tests/project.spec.js [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+                request(app)
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: Again, do not use the server's instance here because w are relying on the URL string. Additionally, remember to return promise from request.

File: server/api/tests/project.spec.js [hunk 10]
```diff
@@ -0,0 +1,353 @@
...
+                .get(`/api/projects/${projectId}`)
+                .expect(200)
+                .then(({body}) => {
+                    let data = body[0];
+                    expect(data.map.name).toBe(mapName);
+                    expect(data.exec).toBe(null);
+                    expect(data.project.name).toBe(projectName)
+                    done();
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: Once all the above will be done you will not need for done.

File: server/api/tests/project.spec.js [hunk 11]
```diff
@@ -0,0 +1,353 @@
...
+
+   
+
+  });
+
+  describe('Negative', () => {
+
+    describe(`POST /`, () => {
+        it(`should respond with a 500 proper msg`, function (done) {
+            try{
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: No need for try/catch if you are not using async/await. What's more, you will have a timeout here because your code will not wait for async execution.

File: server/api/tests/project.spec.js [hunk 12]
```diff
@@ -0,0 +1,353 @@
...
+    });
+
+   
+
+  });
+
+  describe('Negative', () => {
+
+    describe(`POST /`, () => {
+        it(`should respond with a 500 proper msg`, function (done) {
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: Arrow function

File: server/api/tests/project.spec.js [hunk 13]
```diff
@@ -0,0 +1,353 @@
...
+            try{
+                request(app)
+                .post(`/api/projects`)
+                .expect(500)
+                .then(res => {
+                    done();
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: Once all the above will be done you will not need for done.

File: server/api/tests/project.spec.js [hunk 14]
```diff
@@ -0,0 +1,353 @@
...
+   
+
+  });
+
+  describe('Negative', () => {
+
+    describe(`POST /`, () => {
+        it(`should respond with a 500 proper msg`, function (done) {
+            try{
+                request(app)
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: Again, do not use server's instance and do return promise from supertest.

File: server/api/tests/project.spec.js [hunk 15]
```diff
@@ -0,0 +1,353 @@
...
+            catch(err){
+                console.log(err.message);
+                throw err;
+            }
+        }) 
+    });
+
+    describe(`POST /create`, () => {
+        it(`should respond with a 500 proper msg`, function (done) { 
+          try{
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: No need for try/catch if you are not using async/await. What's more, you will have a timeout here because your code will not wait for async execution.

File: server/api/tests/project.spec.js [hunk 16]
```diff
@@ -0,0 +1,353 @@
...
+            }
+            catch(err){
+                console.log(err.message);
+                throw err;
+            }
+        }) 
+    });
+
+    describe(`POST /create`, () => {
+        it(`should respond with a 500 proper msg`, function (done) {
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: Arrow function

File: server/api/tests/project.spec.js [hunk 17]
```diff
@@ -0,0 +1,353 @@
...
+
+    describe(`POST /create`, () => {
+        it(`should respond with a 500 proper msg`, function (done) { 
+          try{
+              request(app)
+              .post(`/api/projects/create`)
+              .send({})
+              .expect(500)
+              .then(res => {
+                  done();
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: No need for done

File: server/api/tests/project.spec.js [hunk 18]
```diff
@@ -0,0 +1,353 @@
...
+                console.log(err.message);
+                throw err;
+            }
+        }) 
+    });
+
+    describe(`POST /create`, () => {
+        it(`should respond with a 500 proper msg`, function (done) { 
+          try{
+              request(app)
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: Do not use server's instance, return promise returned by supertest

File: server/api/tests/project.spec.js [hunk 19]
```diff
@@ -0,0 +1,353 @@
...
+          }
+          catch(err){
+              console.log(err.message);
+              throw err;
+          }
+        });
+      });
+
+    describe(`GET /:id/detail`, () => {
+        it(`should respond with a 500 proper msg`, function (done) {
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: Arrow function

File: server/api/tests/project.spec.js [hunk 20]
```diff
@@ -0,0 +1,353 @@
...
+              console.log(err.message);
+              throw err;
+          }
+        });
+      });
+
+    describe(`GET /:id/detail`, () => {
+        it(`should respond with a 500 proper msg`, function (done) { 
+            try{
+                request(app)
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: Do not use server's instance, return promised coming from supertest

File: server/api/tests/project.spec.js [hunk 21]
```diff
@@ -0,0 +1,353 @@
...
+          catch(err){
+              console.log(err.message);
+              throw err;
+          }
+        });
+      });
+
+    describe(`GET /:id/detail`, () => {
+        it(`should respond with a 500 proper msg`, function (done) { 
+            try{
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: No need for try/catch if you are not using async/await. What's more, you will have a timeout here because your code will not wait for async execution.

File: server/api/tests/project.spec.js [hunk 22]
```diff
@@ -0,0 +1,353 @@
...
+                request(app)
+                .get(`/api/projects/0/detail`)
+                .expect(404)
+                .then(res => {              
+                    done();
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: No need for done

File: server/api/tests/project.spec.js [hunk 23]
```diff
@@ -0,0 +1,353 @@
...
+            }
+            catch(err){
+                console.log(err.message);
+                throw err;
+            }
+        });
+    });
+
+    describe(`PUT /:id/update`, () => {
+        it(`should respond with a 500 proper msg`, function (done) {
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: Arrow function

File: server/api/tests/project.spec.js [hunk 24]
```diff
@@ -0,0 +1,353 @@
...
+            let newDescription = 'simple description';
+            try{
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: No need for try/catch if you are not using async/await. What's more, you will have a timeout here because your code will not wait for async execution.

File: server/api/tests/project.spec.js [hunk 25]
```diff
@@ -0,0 +1,353 @@
...
+                request(app)
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: Return promise coming from supertest, use server's address instead of server's instance.

File: server/api/tests/project.spec.js [hunk 26]
```diff
@@ -0,0 +1,353 @@
...
+                .put(`/api/projects/0/update`)
+                .send({description:newDescription})
+                .expect(500)
+                .then(res => {   
+                    done();
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: No need for done

File: server/api/tests/project.spec.js [hunk 27]
```diff
@@ -0,0 +1,353 @@
...
+            }
+            catch(err){
+                console.log(err.message);
+                throw err;
+            }
+        });
+    });
+
+    describe(`DELETE /:jobId/ `, () => {
+        it(`should respond with a 500 proper msg`, function (done) {
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: Arrow function

File: server/api/tests/project.spec.js [hunk 28]
```diff
@@ -0,0 +1,353 @@
...
+            try{
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: No need for try/catch if you are not using async/await. What's more, you will have a timeout here because your code will not wait for async execution.

File: server/api/tests/project.spec.js [hunk 29]
```diff
@@ -0,0 +1,353 @@
...
+                request(app)
+                .delete(`/api/projects/0/delete`)
+                .expect(500)
+                .then(res => {
+                    done();
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: No need for done

File: server/api/tests/project.spec.js [hunk 30]
```diff
@@ -0,0 +1,353 @@
...
+                console.log(err.message);
+                throw err;
+            }
+        });
+    });
+
+    describe(`DELETE /:jobId/ `, () => {
+        it(`should respond with a 500 proper msg`, function (done) {
+            try{
+                request(app)
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: Return it, do not use server's instance, use server's url.

File: server/api/tests/project.spec.js [hunk 31]
```diff
@@ -0,0 +1,353 @@
...
+            }
+            catch(err){
+                console.log(err.message);
+                throw err;
+            }
+        });
+    });
+
+    describe(`GET /:projectId/ `, () => {
+        it(`should respond with a 500 proper msg`, function (done) {
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: Arrow function

File: server/api/tests/project.spec.js [hunk 32]
```diff
@@ -0,0 +1,353 @@
...
+            try{
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: No need for try/catch if you are not using async/await. What's more, you will have a timeout here because your code will not wait for async execution.

File: server/api/tests/project.spec.js [hunk 33]
```diff
@@ -0,0 +1,353 @@
...
+                request(app)
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: Return it, do not use server's instance

File: server/api/tests/project.spec.js [hunk 34]
```diff
@@ -0,0 +1,353 @@
...
+                .get(`/api/projects/0`)
+                .expect(500)
+                .then(() => {
+                    done();
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: No need if above is fulfilled

File: server/app.js [hunk 35]
```diff
@@ -124,4 +124,6 @@ app.io = io;
 server.listen(port, () => {
     winston.log('info', `Running on localhost:${port}`);
     bootstrap(app);
-});
\ No newline at end of file
+});
+
+module.exports = app;
```
Comment (ChBernat) [inline server/app.js]: Remove it, we agreed to use server's url instead of the instance for now. I will prepare changes that will allow to run server before all the tests.

File: server/package.json [hunk 36]
```diff
@@ -32,12 +32,21 @@
...
     "winston": "3.1.0",
     "winston-mongodb": "4.0.3"
   },
   "devDependencies": {
-    "speculate": "1.7.4"
+    "@types/jest": "^24.0.18",
+    "@types/supertest": "^2.0.8",
+    "speculate": "1.7.4",
+    "supertest": "^4.0.2",
+    "axios": "^0.19.0",
```
Comment (ChBernat) [inline server/package.json]: Remove it, we agreed not to use axios

File: server/package.json [hunk 37]
```diff
@@ -32,12 +32,21 @@
...
+    "chance": "^1.1.0",
+    "jest": "^24.9.0",
+    "json-schema-faker": "^0.5.0-rc17",
+    "rosie": "^2.0.1"
```
Comment (ChBernat) [inline server/package.json]: Remove it, we are not using rosie

File: server/api/tests/project.spec.js [hunk 38]
```diff
@@ -0,0 +1,353 @@
...
+            }
+        });
+    });
+
+    describe(`PUT /:id/update`, () => {
+        it(`should respond with the updated project`, function (done) {
+            let newDescription = 'simple description';
+            let projectName = fixedProjects[0].name;
+            let projectId = testDataManager.getProjectIdByName(projectName)
+            try{
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: No need for try/catch if you are not using async/await. What's more, you will have a timeout here because your code will not wait for async execution.

File: server/api/tests/project.spec.js [hunk 39]
```diff
@@ -0,0 +1,353 @@
...
+                request(app)
+                .put(`/api/projects/${projectId}/update`)
+                .send({description:newDescription})
+                .expect(200)
+                .then(res => {   
+                    testDataManager.updateCollection(projectId,res.body)           
+                    expect(res.body.name).toEqual(projectName);
+                    expect(res.body.description).toEqual(newDescription);
+                    done();
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: Once all the above will be done you will not need for done.

File: server/api/tests/project.spec.js [hunk 40]
```diff
@@ -0,0 +1,353 @@
...
+        });
+    });
+
+    describe(`PUT /:id/update`, () => {
+        it(`should respond with the updated project`, function (done) {
+            let newDescription = 'simple description';
+            let projectName = fixedProjects[0].name;
+            let projectId = testDataManager.getProjectIdByName(projectName)
+            try{
+                request(app)
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: Supertest returns a promise, return it and do not use server's instance. Use url instead.

File: server/api/tests/project.spec.js [hunk 41]
```diff
@@ -0,0 +1,353 @@
...
+            }
+            catch(err){
+                console.log(err.message);
+                throw err;
+            }
+        });
+    });
+
+    describe(`PUT /:id/update`, () => {
+        it(`should respond with the updated project`, function (done) {
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: Arrow function

File: server/api/tests/project.spec.js [hunk 42]
```diff
@@ -0,0 +1,353 @@
...
+            throw err;
+        }
+      });
+    });
+
+    describe(`GET /:id/detail`, () => {
+        it(`should respond with the specific project`, function (done) { 
+            let projectName = fixedProjects[0].name;
+            let projectId = testDataManager.getProjectIdByName(projectName)
+            try{
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: No need for try/catch if you are not using async/await. What's more, you will have a timeout here because your code will not wait for async execution.

File: server/api/tests/project.spec.js [hunk 43]
```diff
@@ -0,0 +1,353 @@
...
+            throw err;
+        }
+      })
+    });
+   
+
+    describe(`POST /create`, () => {
+      it(`should respond with the new project`, function (done) { 
+        let randomData = testDataManager.generateNewItem()
+        try{
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: No need for try/catch if you are not using async/await. What's more, you will have a timeout here because your code will not wait for async execution.

File: server/api/tests/project.spec.js [hunk 44]
```diff
@@ -0,0 +1,353 @@
...
+            throw err;
+        }
+      })
+
+      //adding some parameter to the api like page, sort and globalFilter
+      it(`should respond with a list of projects`, function (done) {
+        let globalFilter = fixedProjects[0].name;
+        let sort = 'name';
+        let page = '1';
+        try{
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: No need for try/catch if you are not using async/await. What's more, you will have a timeout here because your code will not wait for async execution.

File: server/api/tests/project.spec.js [hunk 45]
```diff
@@ -0,0 +1,353 @@
...
+
+
+
+  describe('Positive', () => {
+
+
+    describe(`POST /`, () => {
+      it(`should respond with a list of projects`, function (done) {
+        let randomIndex = Math.floor(Math.random() * testDataManager.collection.length);  
+        try{
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: No need for try/catch if you are not using async/await. What's more, you will have a timeout here because your code will not wait for async execution.

File: server/api/tests/project.spec.js [hunk 46]
```diff
@@ -0,0 +1,353 @@
...
+        }
+        catch(err){
+            console.log(err.message);
+            throw err;
+        }
+      });
+    });
+
+    describe(`GET /:id/detail`, () => {
+        it(`should respond with the specific project`, function (done) {
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: Arrow function

File: server/api/tests/project.spec.js [hunk 47]
```diff
@@ -0,0 +1,353 @@
...
+        catch(err){
+            console.log(err.message);
+            throw err;
+        }
+      })
+    });
+   
+
+    describe(`POST /create`, () => {
+      it(`should respond with the new project`, function (done) {
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: Arrow function

File: server/api/tests/project.spec.js [hunk 48]
```diff
@@ -0,0 +1,353 @@
...
+            }); 
+        }
+        catch(err){
+            console.log(err.message);
+            throw err;
+        }
+      })
+
+      //adding some parameter to the api like page, sort and globalFilter
+      it(`should respond with a list of projects`, function (done) {
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: Arrow function

File: server/api/tests/project.spec.js [hunk 49]
```diff
@@ -0,0 +1,353 @@
...
+        done();
+    })
+
+
+
+  describe('Positive', () => {
+
+
+    describe(`POST /`, () => {
+      it(`should respond with a list of projects`, function (done) {
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: Arrow function

File: server/api/tests/project.spec.js [hunk 50]
```diff
@@ -0,0 +1,353 @@
...
+
+describe('Projects e2e tests', () => {
+
+    beforeEach(async (done) => {
+        fixedProjects =  [
+            {name:'project 1'}, 
+            {name:'project 2'},
+            {name:'project 3'},
+        ]
+        await testDataManager.clear(ProjectModel);
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: No longer needed, take a look at vaults example.

File: server/api/tests/project.spec.js [hunk 51]
```diff
@@ -0,0 +1,353 @@
+const {initTestDataManager, generateProjects} = require('./factories');
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: initTestDataManager no long exists, you have to import TestDataManager class and instantiate it yourself.

File: server/api/tests/project.spec.js [hunk 52]
```diff
@@ -0,0 +1,353 @@
...
+
+
+  describe('Positive', () => {
+
+
+    describe(`POST /`, () => {
+      it(`should respond with a list of projects`, function (done) {
+        let randomIndex = Math.floor(Math.random() * testDataManager.collection.length);  
+        try{
+            request(app)
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: Supertest returns a promise, return it and do not use server's instance. Use url instead.

File: server/api/tests/project.spec.js [hunk 53]
```diff
@@ -0,0 +1,353 @@
...
+        }
+      })
+
+      //adding some parameter to the api like page, sort and globalFilter
+      it(`should respond with a list of projects`, function (done) {
+        let globalFilter = fixedProjects[0].name;
+        let sort = 'name';
+        let page = '1';
+        try{
+            request(app)
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: Supertest returns a promise, return it and do not use server's instance. Use url instead.

File: server/api/tests/project.spec.js [hunk 54]
```diff
@@ -0,0 +1,353 @@
...
+            {name:'project 3'},
+        ]
+        await testDataManager.clear(ProjectModel);
+        await testDataManager.initialise(
+            generateProjects,
+            null,
+            'name description archived maps',
+            fixedProjects
+        );
+        done();
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: Once all the above will be done you will not need for done.

File: server/api/tests/project.spec.js [hunk 55]
```diff
@@ -0,0 +1,353 @@
...
+let fixedProjects;
+let mapId;
+
+
+
+setupDB('test');
+
+const testDataManager = initTestDataManager(ProjectModel);
+
+function createMap(projectId,index,mapName){
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: Are you testing the create map / map endpoints in this file? If not then please, use json-schema-faker to generate the proper factory and use TestDataManager to initialize the collection of maps in beforeAll/beforeEach.

File: server/api/tests/project.spec.js [hunk 56]
```diff
@@ -0,0 +1,353 @@
...
+    return new Promise((resolve,reject) => {
+        request(app)
+        .post(`/api/maps/create`)
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: You are not testing maps in this file, use factory and TestDataManager instead!

File: server/api/tests/project.spec.js [hunk 57]
```diff
@@ -0,0 +1,353 @@
...
+        try{
+            request(app)
+            .post(`/api/projects`)
+            .send({options:{}})
+            .expect(200)
+            .then(res => {
+                expect(res.body.items.length).toEqual(testDataManager.collection.length)
+                expect(res.body.totalCount).toEqual(testDataManager.collection.length)
+                expect(res.body.items[randomIndex].id).toEqual(testDataManager.collection[randomIndex].id);
+                done();
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: Once all the above will be done you will not need for done.

File: server/api/tests/project.spec.js [hunk 58]
```diff
@@ -0,0 +1,353 @@
...
+            .post(`/api/projects`)
+            .send({options:{sort:sort,page:page,globalFilter:globalFilter}})
+            .expect(200)
+            .then(res => {
+                expect(res.body.items.length).toEqual(1)
+                expect(res.body.totalCount).toEqual(1)
+                if(res.body.items.length){
+                    expect(res.body.items[0].name).toEqual(fixedProjects[0].name);
+                }
+                done();
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: Once all the above will be done you will not need for done.

File: server/api/tests/project.spec.js [hunk 59]
```diff
@@ -0,0 +1,353 @@
...
+        let randomData = testDataManager.generateNewItem()
+        try{
+            request(app)
+            .post(`/api/projects/create`)
+            .send(randomData)
+            .expect(200)
+            .then(res => {
+                testDataManager.pushToCollection(res.body)
+                expect(res.body.name).toEqual(randomData.name);
+                done();
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: Once all the above will be done you will not need for done.

File: server/api/tests/project.spec.js [hunk 60]
```diff
@@ -0,0 +1,353 @@
...
+        it(`should respond with the specific project`, function (done) { 
+            let projectName = fixedProjects[0].name;
+            let projectId = testDataManager.getProjectIdByName(projectName)
+            try{
+                request(app)
+                .get(`/api/projects/${projectId}/detail`)
+                .expect(200)
+                .then(res => {              
+                    expect(res.body.name).toEqual(projectName);
+                    done();
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: Once all the above will be done you will not need for done.

File: server/api/tests/project.spec.js [hunk 61]
```diff
@@ -0,0 +1,353 @@
+const {initTestDataManager, generateProjects} = require('./factories');
+const ProjectModel = require('../../api/models/project.model');
+const {setupDB} = require('./helpers/test-setup');
+const request = require('supertest');
+const app = require('../../app');
+let fixedProjects;
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: This should not be defined here, in main namespace but rather in a callback to describe.

File: server/api/tests/project.spec.js [hunk 62]
```diff
@@ -0,0 +1,353 @@
...
+let mapId;
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: There's no use of this variable in test file. Please, delete it.

File: server/api/tests/project.spec.js [hunk 63]
```diff
@@ -0,0 +1,353 @@
...
+const testDataManager = initTestDataManager(ProjectModel);
+
+function createMap(projectId,index,mapName){
+    return new Promise((resolve,reject) => {
+        request(app)
+        .post(`/api/maps/create`)
+        .send({ name: mapName,project:projectId })
+        .then(res=> {
+            testDataManager.collection[index].maps.push(res.body.id)
+            mapId = res.body.id;
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: Again, no use for it.

File: server/api/tests/project.spec.js [hunk 64]
```diff
@@ -0,0 +1,353 @@
...
+
+    beforeEach(async (done) => {
+        fixedProjects =  [
+            {name:'project 1'}, 
+            {name:'project 2'},
+            {name:'project 3'},
+        ]
+        await testDataManager.clear(ProjectModel);
+        await testDataManager.initialise(
+            generateProjects,
```
Comment (ChBernat) [inline server/api/tests/project.spec.js]: This should be an invocation now. ThestDataManager's initialise function was changed and requires an object instead.

File: server/api/tests/project.spec.js (hunk 7 above)
Comment (DavidIzaac) [reply server/api/tests/project.spec.js]: This is not part of the test's setup because the projects are created in the beforeEach function. I can't create a map without a project so it's mean that I need to create map also in my beforeEach function. And because I need to create map only for the get :projectId endpoint, so it cannot be a part of the test's setup.

File: server/api/tests/factories/projects.factory.js [hunk 65]
```diff
@@ -0,0 +1,36 @@
...
+};
+
+const arrayProjectSchema = {
+    type: 'array',
+    items: singleProjectSchema,
+    maxItems: 2,
+};
+
+
+function getSimpleMapSchema(name) {
```
Comment (arispen) [inline server/api/tests/factories/projects.factory.js]: shouldn't this be in a separate file?

File: server/api/tests/projects.test.js [hunk 66]
```diff
@@ -0,0 +1,223 @@
...
+                return request(baseApiURL)
+                    .delete(`/projects/${randomProject.id}/delete`)
+                    .expect(200)
+                    .then(res => {
+                        expect('OK');
+                    });
+            });
+        });
+
+        //BUG (KAH -20)
```
Comment (arispen) [inline server/api/tests/projects.test.js]: do we _really_ need this commented code here?

File: server/api/tests/projects.test.js [hunk 67]
```diff
@@ -0,0 +1,223 @@
...
+                    throw err;
+                }
+            });
+        });
+    });
+
+    describe('Negative', () => {
+
+        describe(`POST /`, () => {
+            it(`should respond with a 500 proper msg`, (done) => {
```
Comment (arispen) [inline server/api/tests/projects.test.js]: where is that `proper msg` ?

File: server/api/tests/projects.test.js [hunk 68]
```diff
@@ -0,0 +1,223 @@
...
+    try {
+        const map = await Map.create(generatedMap);
+        await ProjectModel.findByIdAndUpdate({_id: projectId}, {$push: {maps: map.id}}, {new: true});
+        testDataManager.collection[index].maps.push(map.id)
+    } catch (err) {
+        throw err;
+    }
+}
+
+describe('Projects e2e tests', () => {
```
Comment (arispen) [inline server/api/tests/projects.test.js]: 🤔 we should agree on _one_ test naming convention in the nearest future

File: server/api/tests/projects.test.js (hunk 66 above)
Comment (ChBernat) [reply server/api/tests/projects.test.js]: @DavidIzaac ^

File: server/api/tests/projects.test.js (hunk 68 above)
Comment (ChBernat) [reply server/api/tests/projects.test.js]: Please, start a discussion in #general on Slack instead.

File: server/api/tests/projects.test.js (hunk 67 above)
Comment (ChBernat) [reply server/api/tests/projects.test.js]: @DavidIzaac ^

File: server/api/tests/factories/projects.factory.js (hunk 65 above)
Comment (DavidIzaac) [reply server/api/tests/factories/projects.factory.js]: The purpose of this function is to pupulate the maps field of the **project** schema.

File: server/api/tests/projects.test.js (hunk 66 above)
Comment (DavidIzaac) [reply server/api/tests/projects.test.js]: When we will fix the KAH-20 bug this commented code will make sense

File: server/api/tests/factories/projects.factory.js (hunk 65 above)
Comment (theflexbird) [reply server/api/tests/factories/projects.factory.js]: I would recommend changing the name. The project has multiple maps, so for the purpose of the tests, it needs to create some map along with project. When map test will be done it needs take the map schema form `maps.factory.js`. We can also move it there right now and import it in the test itself.

File: server/api/tests/projects.test.js (hunk 68 above)
Comment (theflexbird) [reply server/api/tests/projects.test.js]: Maybe `Projects API tests`

File: server/api/tests/factories/projects.factory.js (hunk 65 above)
Comment (DavidIzaac) [reply server/api/tests/factories/projects.factory.js]: Go for maps.factory
